src/Microsoft.AspNet.Mvc.Core/ViewDataDictionary.cs

```diff
@@ -354,15 +332,24 @@ public ViewDataInfo GetViewDataInfo(string expression)
         protected virtual void SetModel(object value)
         {
             EnsureCompatible(value);
-            _model = value;
 
             // Reset or override ModelMetadata based on runtime value type. Fall back to declared type if value is
             // null. When called from the Model setter, ModelMetadata will (temporarily) be null. When called from
             // a constructor, current ModelMetadata may already be set to preserve customizations made in parent scope.
             var modelType = GetModelType(value);
-            if (ModelMetadata == null || ModelMetadata.ModelType != modelType)
+            if (ModelExplorer?.Metadata.ModelType != modelType)
+            {
+                ModelExplorer = _metadataProvider.GetModelExplorerForType(modelType, value);
+            }
+            else if (object.ReferenceEquals(value, Model))
```

Comment:
It's not needed for correctness. Mainly wanted to document that it can and does happen.